contracts/src/v0.8/keystone/CapabilityRegistry.sol

```diff
@@ -184,6 +217,38 @@ contract CapabilityRegistry is OwnerIsCreator, TypeAndVersionInterface {
     return s_nodeOperators[nodeOperatorId];
   }
 
+  /// @notice Adds nodes. Nodes can be added with deprecated capabilities to
+  /// avoid breaking changes when deprecating capabilities.
+  /// @param nodes The nodes to add
+  function addNodes(Node[] calldata nodes) external {
+    for (uint256 i; i < nodes.length; ++i) {
+      Node memory node = nodes[i];
+
+      if (node.supportedCapabilityIds.length == 0) revert InvalidNodeCapabilities(node.supportedCapabilityIds);
+
+      for (uint256 j; j < node.supportedCapabilityIds.length; ++j) {
+        if (!s_capabilityIds.contains(node.supportedCapabilityIds[j]))
+          revert InvalidNodeCapabilities(node.supportedCapabilityIds);
+      }
+
+      bool nodeExists = s_nodes[node.p2pId].supportedCapabilityIds.length > 0;
+      if (nodeExists || bytes32(node.p2pId) == bytes32("")) revert InvalidNodeP2PId(node.p2pId);
+
+      NodeOperator memory nodeOperator = s_nodeOperators[node.nodeOperatorId];
```

Comment:
Wdyt about moving this check to the beginning of the iteration so we can short circuit the code earlier?